internal/auth/client_kerberos_test.go

```diff
@@ -0,0 +1,355 @@
+		t.Fatal("Failed to write keytab")
+	}
+	config.Kerberos.Keytab = kt.Name()
+	err = kt.Close()
+	if err != nil {
+		t.Fatal("Failed to close keytab file", err)
+	}
+
+	c, err := auth.NewKerberosClient(
+		config,
+		log.NewTestLogger(t),
+		metrics.New(dummy.New()),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return c
+}
+
+func TestKerberosPasswordAuth(t *testing.T) {
+	config := configuration.AuthConfig{
+		Method: configuration.AuthMethodKerberos,
+		Kerberos: configuration.AuthKerberosClientConfig{
+			EnforceUsername: true,
+			AllowPassword: true,
+			ConfigPath: "testdata/krb5.conf",
+		},
+	}
+	c := setupKerberosClient(t, config)
+
+	ctx := c.Password("foo", []byte("test"), "asdf", net.ParseIP("127.0.0.1"))
+	if !ctx.Success() {
+		assert.Fail(t, "Failed to login as foo", ctx.Error())
+	}
+
+	ctx = c.Password("foo", []byte("wrongpass"), "asdf", net.ParseIP("127.0.0.1"))
+	if ctx.Success() {
+		assert.Fail(t, "Logged in as foo with wrong password")
+	}
+}
+
+func TestKerberosNoPassword(t *testing.T) {
+	config := configuration.AuthConfig{
+		Method: configuration.AuthMethodKerberos,
+		Kerberos: configuration.AuthKerberosClientConfig{
+			EnforceUsername: true,
+			AllowPassword: false,
+			ConfigPath: "testdata/krb5.conf",
+		},
+	}
+	c := setupKerberosClient(t, config)
+
+	ctx := c.Password("foo", []byte("test"), "asdf", net.ParseIP("127.0.0.1"))
+	if ctx.Success() {
+		assert.Fail(t, "Logged in as foo even though password is disallowed")
+	}
+
+	ctx = c.Password("foo", []byte("wrongpass"), "asdf", net.ParseIP("127.0.0.1"))
+	if ctx.Success() {
+		assert.Fail(t, "Logged in as foo with wrong password")
+	}
+}
+
+func TestKerberosGSSAPIAuth(t *testing.T) {
+	config := configuration.AuthConfig{}
+	structutils.Defaults(&config)
+	config.Method = configuration.AuthMethodKerberos
+	config.Kerberos.ConfigPath = "testdata/krb5.conf"
+
+	authCl := setupKerberosClient(t, config)
+	userCl := krbAuth(t, "foo", "test")
+
+	gssClient := testGssApiClient{
+		client: userCl,
+	}
+
+
+	name, err := doGssAuth(
+		t,
+		"foo",
+		"host/testing.containerssh.io",
+		false,
+		authCl.GSSAPIConfig(
+			"asdf",
+			net.IPv4(127, 0, 0, 1),
+		),
+		gssClient,
+	)
+	if err != nil {
+		assert.Fail(t, "GSSAPI Failure", err)
+	}
+	if name != "foo" {
+		assert.Fail(t, "Logged in with the wrong name", name)
+	}
+
+	name, err = doGssAuth(
+		t,
+		"foo",
+		"host/testing.containerssh.io",
+		true,
+		authCl.GSSAPIConfig(
+			"asdf",
+			net.IPv4(127, 0, 0, 1),
+		),
+		gssClient,
+	)
+	if err != nil {
+		assert.Fail(t, "GSSAPI Failure", err)
+	}
+	if name != "foo" {
+		assert.Fail(t, "Logged in with the wrong name", name)
+	}
+
+	_, err = doGssAuth(
+		t,
+		"bar",
+		"host/testing.containerssh.io",
+		false,
+		authCl.GSSAPIConfig(
+			"asdf",
+			net.IPv4(127, 0, 0, 1),
+		),
+		gssClient,
+	)
+	if err == nil {
+		assert.Fail(t, "Invalid username was successfully authenticated")
+	}
+}
+
+
+// Partially from kerberosAuthClient::Password
```

Comment:
What is the source of this? Do we need to consider licensing issues here?

Reply:
I meant that it was initially taken from `client_kerberos.go` the password auth routine. Though after many iterations it's basically unrecognizable. Removed the comment.